c10/cuda/CUDAException.cpp

```diff
@@ -26,10 +111,197 @@ void c10_cuda_check_implementation(
   check_message.append("CUDA error: ");
   check_message.append(cudaGetErrorString(err));
   check_message.append(c10::cuda::get_cuda_check_suffix());
+  check_message.append("\n");
+  if (include_device_assertions) {
+    check_message.append(c10_retrieve_device_side_assertion_info());
+  }
 #endif
 
   TORCH_CHECK(false, check_message);
 }
 
+CUDAKernelLaunchRegistry::CUDAKernelLaunchRegistry()
+    : do_all_devices_support_managed_memory(
+          check_if_all_devices_support_managed_memory()) {
+  for (C10_UNUSED const auto _ : c10::irange(get_device_count())) {
+    uvm_assertions.emplace_back(nullptr, uvm_deleter);
+  }
+
+  kernel_launches.resize(max_kernel_launches);
+}
+
+uint32_t CUDAKernelLaunchRegistry::insert(
+    const std::string& launch_filename,
+    const std::string& launch_function,
+    const uint32_t launch_linenum,
+    const std::string& kernel_name,
+    const int32_t stream_id) {
+#ifdef TORCH_USE_CUDA_DSA
+#ifdef TORCH_CUDA_DSA_WITH_BACKTRACE
```

Comment:
Repeating the comment: can we make stack control a runtime opt-in instead? It doesn't have any static impact on the performance. It can be an env variable perhaps.

Also as Ed mentioned if there's little perf impact from enabling the logging by default - we should make `TORCH_USE_CUDA_DSA` a runtime option too.

Reply:
I have added runtime options for both stacktrace acquisition and DSA as a whole.